### PR TITLE
refactor(provider/amazon): Separate load balancer types

### PR DESCRIPTION
**Pull request by jrsquared**
A first pass at separating out load balancer types. There is still some confusion around what gets transformed after the load balancers get read, and the different data structures when load balancers are retrieved as group vs a single load balancer, but this gets us a LOT closer to actual valid shapes.

### CODE REVIEW

File: app/scripts/modules/amazon/src/domain/IAmazonLoadBalancer.ts [hunk 1]
```diff
@@ -76,14 +78,88 @@ export interface ITargetGroup {
...
+  type: string; // returned from clouddriver
   vpcId?: string;
   vpcName?: string;
 }
+
+export interface IUpsertAmazonLoadBalancerCommand extends IUpsertLoadBalancerCommand {
+  availabilityZones: { [region: string]: string[] };
+  isInternal: boolean;
+  // If loadBalancerType is not provided, will default to 'classic' for bwc
+  loadBalancerType?: 'classic' | 'application' | 'network';
```
Comment (anotherchrisberry) [inline app/scripts/modules/amazon/src/domain/IAmazonLoadBalancer.ts]: Is `network` future-proofing or a real thing?

File: app/scripts/modules/amazon/src/domain/IAmazonLoadBalancerSourceData.ts [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+  detachedInstances: string[];
+  instances: IAmazonLoadBalancerInstanceSourceData[];
+  isDisabled: boolean;
+  name: string;
+  region: string;
+}
+
+export interface IAmazonInstanceHealthSourceData {
+  type: string;
+  state: 'InService' | 'OutOfService' | 'Unknown' | 'healthy';
```
Comment (anotherchrisberry) [inline app/scripts/modules/amazon/src/domain/IAmazonLoadBalancerSourceData.ts]: Should `healthy` be capitalized? If not, might be worth adding a comment to explain the discrepancy.

File: app/scripts/modules/amazon/src/domain/IAmazonLoadBalancerSourceData.ts [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+  reasonCode: 'ELB' | 'Instance' | 'N/A';
+  description: string;
+}
+
+export interface IClassicListenerSourceData {
+  instancePort: number;
+  instanceProtocol: string;
+  loadBalancerPort: number;
+  protocol: string;
+};
```
Comment (anotherchrisberry) [inline app/scripts/modules/amazon/src/domain/IAmazonLoadBalancerSourceData.ts]: semicolon

File: app/scripts/modules/amazon/src/domain/IAmazonLoadBalancerSourceData.ts [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+  name: string;
+  region: string;
+  scheme: 'internal' | 'internet-facing';
+  securityGroups: string[];
+  serverGroups: IAmazonLoadBalancerServerGroupSourceData[];
+  subnets: string[];
+  type: string;
+  vpcId: string;
+  vpcid?: string;
+};
```
Comment (anotherchrisberry) [inline app/scripts/modules/amazon/src/domain/IAmazonLoadBalancerSourceData.ts]: semicolon

File: app/scripts/modules/amazon/src/domain/IAmazonLoadBalancerSourceData.ts [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+  loadBalancerName: string;
+  name: string;
+  region: string;
+  scheme: 'internal' | 'internet-facing';
+  securityGroups: string[];
+  serverGroups: IAmazonLoadBalancerServerGroupSourceData[];
+  subnets: string[];
+  type: string;
+  vpcId: string;
+  vpcid?: string;
```
Comment (anotherchrisberry) [inline app/scripts/modules/amazon/src/domain/IAmazonLoadBalancerSourceData.ts]: worth a comment explaining why there are both `vpcId` and `vpcid` fields

File: app/scripts/modules/amazon/src/domain/IAmazonLoadBalancerSourceData.ts [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+  vpcId: string;
+}
+
+export interface IApplicationLoadBalancerSourceData extends IAmazonLoadBalancerSourceData {
+  ipAddressType: 'ipv4' | 'dualstack';
+  listeners: {
+    defaultActions: {
+      targetGroupName: string;
+      type: 'forward';
+    }[];
```
Comment (anotherchrisberry) [inline app/scripts/modules/amazon/src/domain/IAmazonLoadBalancerSourceData.ts]: I like the nested declaration approach (vs. declaring a bunch of extra interfaces that are only used in one place and never referenced).

File: app/scripts/modules/amazon/src/domain/IAmazonLoadBalancerSourceData.ts [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    protocol: string;
+  }[];
+  loadBalancerArn: string;
+  loadBalancerType: 'application' | 'network';
+  state: {
+    code: 'active' | 'provisioning' | 'failed';
+    reason?: string;
+  };
+  targetGroups: IAmazonTargetGroupSourceData[];
+};
```
Comment (anotherchrisberry) [inline app/scripts/modules/amazon/src/domain/IAmazonLoadBalancerSourceData.ts]: ;

File: app/scripts/modules/amazon/src/loadBalancer/loadBalancer.transformer.ts [hunk 8]
```diff
@@ -209,6 +225,7 @@ export class AwsLoadBalancerTransformer {
               const certArnParts = cert.certificateArn.split(':');
               const certParts = certArnParts[5].split('/');
               certificates.push({
+                certificateArn: cert.certifcateArn,
```
Comment (anotherchrisberry) [inline app/scripts/modules/amazon/src/loadBalancer/loadBalancer.transformer.ts]: `cert.certificateArn`

File: app/scripts/modules/core/src/domain/ILoadBalancer.ts [hunk 9]
```diff
@@ -31,3 +38,12 @@ export interface ILoadBalancerGroup {
   subgroups?: ILoadBalancerGroup[];
   searchField?: string;
 }
+
+export interface IUpsertLoadBalancerCommand {
+  credentials: string;
+  detail?: string;
+  editMode?: boolean; // Is this even used anymore?
```
Comment (anotherchrisberry) [inline app/scripts/modules/core/src/domain/ILoadBalancer.ts]: only you would know if it's used anymore

File: app/scripts/modules/core/src/loadBalancer/loadBalancer.read.service.ts [hunk 10]
```diff
@@ -22,15 +22,15 @@ export class LoadBalancerReader {
     'ngInject';
   }
 
-  public loadLoadBalancers(applicationName: string): ng.IPromise<ILoadBalancer[]> {
+  public loadLoadBalancers(applicationName: string): ng.IPromise<ILoadBalancerSourceData[]> {
```
Comment (anotherchrisberry) [inline app/scripts/modules/core/src/loadBalancer/loadBalancer.read.service.ts]: while you're in here, might as well remove the `ng.` package

**Comment by anotherchrisberry** [comment]
a few little things; LGTM otherwise

File: app/scripts/modules/amazon/src/domain/IAmazonLoadBalancer.ts (hunk 1 above)
Comment (jrsquared) [inline app/scripts/modules/amazon/src/domain/IAmazonLoadBalancer.ts]: Both.

File: app/scripts/modules/amazon/src/domain/IAmazonLoadBalancerSourceData.ts (hunk 2 above)
Comment (jrsquared) [inline app/scripts/modules/amazon/src/domain/IAmazonLoadBalancerSourceData.ts]: It should not, but it is actually a different health state, so I'll extract this a little further.

File: app/scripts/modules/core/src/domain/ILoadBalancer.ts (hunk 9 above)
Comment (jrsquared) [inline app/scripts/modules/core/src/domain/ILoadBalancer.ts]: one way to find out...
fwiw, full text search across orca and clouddriver, doesn't show up in clouddriver, and only shows up in a test in orca.

**Pull request merged**